### PR TITLE
feat: Redesign batch execution with pluggable schedulers

### LINKED ISSUE #347: Split out batch executor
Currently, batch execution happens inside the batches API by creating an executor for each input. We should probably split this out into a separate batch executor that intelligently manages this. We may want to consider work queue, etc.

> Would be good to split out the batch execution logic into a separate executor so we can more intelligently manage things.

_Originally posted by @bjchambers in https://github.com/stepflow-ai/stepflow/pull/345#discussion_r2403704324_

### PR DESCRIPTION
**Pull request by bjchambers**
This PR introduces a new execution architecture that enables cross-item optimization and pluggable scheduling strategies for batch workflows.

## New Architecture

- **FlowExecutor**: Unified executor for single and multi-item runs with async task execution using FuturesUnordered
- **DebugExecutor**: Step-by-step debugging with queue/eval API and support for resuming from persistent state
- **ItemsState**: Cross-item coordinator tracking execution state across all items in a batch with O(1) completion checking
- **ItemState**: Per-item state with BitSet-based dependency tracking

## Pluggable Schedulers

- **Scheduler trait**: Interface for custom execution strategies
- **DepthFirstScheduler**: Prioritizes completing items sequentially (minimizes time-to-first-result)
- **BreadthFirstScheduler**: Groups tasks by step across items (maximizes throughput, enables future batch component invocation)

## Key Improvements

- Schedulers query ItemsState for task status instead of maintaining internal tracking (single source of truth for resume/recovery)
- Atomic counter for O(1) is_all_complete() check in execution loop
- Shared test utilities in testing.rs module
- Comprehensive test coverage including edge cases and failure isolation

## Removed

- Legacy WorkflowExecutor and ExecutionTracker
- Redundant per-item spawning with semaphore-based concurrency

Closes #347

### CODE REVIEW

File: stepflow-rs/crates/stepflow-execution/src/debug_executor.rs [hunk 1]
```diff
@@ -0,0 +1,1177 @@
...
+
+        // Create ItemState with workflow input and variables
+        let state = ItemState::new(
+            flow.clone(),
+            input,
+            Arc::new(variables.unwrap_or_default()),
+        );
+
+        // Create workflow-aware execution context
+        let context = ExecutionContext::for_workflow_with_flow(
```
Comment (bjchambers) [inline stepflow-rs/crates/stepflow-execution/src/debug_executor.rs]: Naming? `for_workflow_with_flow`? Is this still needed? Can it be cleaned up after #472?

File: stepflow-rs/crates/stepflow-execution/src/debug_executor.rs [hunk 2]
```diff
@@ -0,0 +1,1177 @@
...
+        // Queue output dependencies
+        let output_needs = self.flow.output().needed_steps(&self.state);
+        for idx in output_needs.iter() {
+            self.state.add_or_update_needed(idx);
+        }
+
+        // Run until output is resolvable and all needed steps complete
+        loop {
+            // Check if output is ready
+            let output_needs = self.flow.output().needed_steps(&self.state);
```
Comment (bjchambers) [inline stepflow-rs/crates/stepflow-execution/src/debug_executor.rs]: It feels like a lot of this debug executor is duplicative. I think we may be able to simplify. It seems like the main thing is that while normal execution is "drive everything to compute the output", this is more of a "drive everything to complete what the user has requested". This makes me wonder if we should have the debug executor manage a separate set of "requested" computations, and use that similar to how the `FlowExecutor` drives things. Then all of these are different forms of `drain_requested(...)`. It may not even need to be a queue -- we may be able to just have a method `eval_requested(...)` and then have the different `eval` methods call it.

File: stepflow-rs/crates/stepflow-execution/src/debug_executor.rs [hunk 3]
```diff
@@ -0,0 +1,1177 @@
...
+    // ========================================================================
+
+    /// Recover execution state from the state store and fix any missed status transitions.
+    ///
+    /// This is called internally by [`resume()`](Self::resume) to restore completed steps
+    /// and fix any status mismatches.
+    ///
+    /// # Returns
+    /// The number of step status corrections made during recovery
+    async fn recover_from_state_store(&mut self) -> Result<usize> {
```
Comment (bjchambers) [inline stepflow-rs/crates/stepflow-execution/src/debug_executor.rs]: This should be two parts -- recreating the `DebugState` from the state store, and then continuing the execution as needed.

File: stepflow-rs/crates/stepflow-execution/src/executor.rs [hunk 4]
```diff
@@ -232,96 +184,78 @@ impl Context for StepflowExecutor {
...
                     } else {
                         stepflow_core::status::ExecutionStatus::Completed
                     };
 
-                    // Update the run status (results already recorded individually)
-                    let _ = executor_clone
-                        .state_store()
+                    // Update the run status
+                    let _ = state_store_clone
                         .update_run_status(run_id, final_status)
```
Comment (bjchambers) [inline stepflow-rs/crates/stepflow-execution/src/executor.rs]: Updating the run status on completion should be part of the FlowExecutor.

File: stepflow-rs/crates/stepflow-execution/src/flow_executor.rs [hunk 5]
```diff
@@ -0,0 +1,871 @@
...
+        }
+    }
+
+    /// Execute all items to completion.
+    ///
+    /// Returns a vector of FlowResults, one per item, in item index order.
+    /// Results are recorded to the state store as items complete.
+    ///
+    /// Tasks are executed concurrently up to `max_in_flight` at a time.
+    pub async fn execute_to_completion(&mut self) -> Result<Vec<FlowResult>> {
```
Comment (bjchambers) [inline stepflow-rs/crates/stepflow-execution/src/flow_executor.rs]: We could allow this to be a stream of results and include the`ItemIndex` in the result. Then, we wouldn't need the order to match, nor would we need to collect all the results.
